### PR TITLE
Fix dynamic graph sizes in message passing exports

### PR DESCRIPTION
**Pull request by thangldw**
## Bug report

Addresses the dynamic graph-size portion of #1335.

Message-passing layers allocate scatter-reduction outputs with `len(bmg.V)`. During `torch.export`, that Python length is specialized to the number of atoms in the export example. Running the exported model on a molecular graph with a different atom count then fails with an out-of-bounds scatter or a dynamic-shape constraint violation.

This change uses the symbolic tensor dimension `bmg.V.shape[0]` for those allocations. The same specialization was present in the shared bond/atom mixins and in both standard and Mol-Atom-Bond final aggregation paths, so all four paths are updated consistently.

`MeanAggregation` is intentionally unchanged, following the performance concern discussed in #1335.

## Example

The regression test exports each message-passing implementation using a three-atom chain and executes the exported program using a five-atom chain. Before this patch, export rejects the atom dimension as specialized to `3`; with this patch, the exported and eager outputs match for:

- `AtomMessagePassing`
- `BondMessagePassing`
- `MABAtomMessagePassing`
- `MABBondMessagePassing`

A separate end-to-end check through ONNX Runtime, using `BondMessagePassing` and `SumAggregation`, changed from an out-of-bounds `ScatterElements` failure to successful inference with a maximum PyTorch/ONNX absolute difference of `4.47e-08`.

## Questions

None. This patch is limited to preserving dynamic graph dimensions and adds no public API, checkpoint, or user-facing behavior changes.

## Checklist

- [x] `black --check`, `isort --check-only`, and `flake8` pass on all changed files
- [x] Dynamic export regression test passes on PyTorch 2.7.0 and 2.13.0
- [x] 128 related unit and integration tests pass on PyTorch 2.7.0
- [x] Appropriate unit tests added

### CODE REVIEW

File: tests/unit/nn/test_message_passing.py
Comment (JacksonBurns) [inline tests/unit/nn/test_message_passing.py:1]: Why do we test on a Mock of `BatchMolGraph` here? Seems like it would make more sense to just the actual class?

And another test is needed, specifically an integration test trying to `export` the entire `MPNN` and comparing inference before and after.

**Review by JacksonBurns — CHANGES_REQUESTED** [review]
Thanks for getting this started, some notes on the tests.

File: tests/unit/nn/test_message_passing.py
Comment (thangldw) [reply tests/unit/nn/test_message_passing.py:1]: Thanks — agreed. I replaced the MockBatchMolGraph with the actual BatchMolGraph, constructed from real MolGraph inputs, and added a shared pytree-registration fixture for export tests. I also added an integration test that builds the full MPNN (BondMessagePassing + SumAggregation + RegressionFFN), records eager inference on a graph with different atom/edge counts, exports the model, and compares inference from the exported module. Validation: 5 targeted export tests pass; 146 NN/model/integration tests pass (3 skipped); black, flake8, isort, and git diff checks pass. I also manually verified the full ONNX Runtime path from 2 molecules / 3 atoms / 2 edges at export to 5 molecules / 19 atoms / 30 edges at inference (max abs diff 4.47e-08).

File: tests/integration/test_export.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+from chemprop.models import MPNN
+from chemprop.nn import BondMessagePassing, RegressionFFN, SumAggregation
+
+
+def make_batch(smiles: list[str]) -> BatchMolGraph:
+    dataset = MoleculeDataset([MoleculeDatapoint.from_smi(smi) for smi in smiles])
+    return collate_batch(dataset)[0]
+
+
+def test_mpnn_export_dynamic_graph_sizes(batch_mol_graph_pytree):
```
Comment (KnathanM) [inline tests/integration/test_export.py]: It looks like `batch_mol_graph_pytree` is passed in but not used? Both here and in `test_message_passing_export_dynamic_graph_sizes`.

File: tests/integration/test_export.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    inference_graph = make_batch(["c1ccccc1", "CCCCCC"])
+    assert export_graph.V.shape[0] != inference_graph.V.shape[0]
+    assert export_graph.E.shape[0] != inference_graph.E.shape[0]
+
+    message_passing = BondMessagePassing()
+    model = MPNN(
+        message_passing, SumAggregation(), RegressionFFN(input_dim=message_passing.output_dim)
+    ).eval()
+    num_atoms = torch.export.Dim("num_atoms", min=1)
+    num_edges = torch.export.Dim("num_edges", min=1)
```
Comment (KnathanM) [inline tests/integration/test_export.py]: Do you need to set `min=1`? Or what would happen if you left it off? It is possible in general to have a batched graph where the number of edges is 0: `make_batch(["C", "S", "N", "O"])`, though this is unlikely in practice. And this line of code is just in the tests and not core code, so doesn't really need to handle all cases.

**Review by KnathanM — COMMENTED** [review]
Thanks for the change. I see it matches what pytorch also [suggests](https://github.com/pytorch/pytorch/blob/7e6bc2fdc12dd310571468a00bbee27f3b9e0bec/torch/_tensor.py#L1196-L1197).

Just a couple comments on the tests and this should be good to merge.

File: tests/integration/test_export.py (hunk 1 above)
Comment (thangldw) [reply tests/integration/test_export.py]: Good catch. The fixture is used for its side effect: it registers BatchMolGraph as a pytree before torch.export runs. I replaced the apparently-unused function arguments in both tests with @pytest.mark.usefixtures("batch_mol_graph_pytree"), which makes that intent explicit.

File: tests/integration/test_export.py (hunk 2 above)
Comment (thangldw) [reply tests/integration/test_export.py]: No, min=1 was not necessary, so I removed the explicit lower bounds. I also changed the full-MPNN inference batch to the exact zero-edge case make_batch(["C", "S", "N", "O"]) and asserted that E.shape[0] == 0. Exported-module inference still matches eager inference. I additionally checked the ONNX Runtime path with 12 edges at export and 0 edges at inference; it passed with a max absolute difference of 3.73e-09.

**Review by KnathanM — APPROVED** [review]
@JacksonBurns Do you think this is ready to merge?

**Review by JacksonBurns — APPROVED** [review]
